### ISSUE TITLE
Unable to obtain list of significant genes for each GO term in GO enrichment test 

### ISSUE THREAD

**Issue by ayeTown** [pantherdb/pantherapi-pyclient#8]
Hi there,
I would like to obtain the list of genes that are in each enriched GO term. It doesn't seem as though the results or even the IDs of the significant genes are provided in the results table, but only the number of genes in the GO term reference list. Is there a way I can obtain these using the API?
Thanks for your time,
AT

**Comment by mugitty** [comment]
Currently the API does not output the genes associated with each term. However, you can obtain the associated genes for each term by running the analysis from the website and exporting the results to JSON or XML.

1. You would start from http://pantherdb.org/
2. Click on the browse button to select a file with the gene ids from the comparison list
3. Select the organism
4. Select Statistical enrichment test and the annotation data set (Example Full GO BP, PANTHER MF, Reactome...)
5. Click submit
6. The system will prompt you to select a reference gene list.  Here you can upload a custom gene list on the right hand side or select a gene list with all genes for the organism
7. The system will now prompt with a selection summary.  Here you can add more comparison lists and also update the  analysis type and correction.
8. Clicking on Launch Analysis will result in running the analysis.
9. From here you can export the results as a table or to get associated input genes use the XML or JSON option.

**Comment by ayeTown** [comment]
Okay, thanks for the instructions and quick response. I think that would be a nice new feature to have. I am using the API within an R Shiny application and making the parameters for the request interactive for the user. So I would have to run it manually for every possible combination which is not feasible.

**Comment by ayeTown** [comment]
@mugitty I have just noticed there is a POST request "geneinfo" which returns associated GO terms with a given gene. Could I use this with the same gene list I input for the GO enrichment and take the intersection of the GO terms to get the associated genes? Only issue I see with this is that there is no gene ID in the results column but a "persistent_id". Can I assume this new ID has the same order as the input gene list?

**Comment by mugitty** [comment]
@ayeTown the "geneinfo" POST request does not return the results in the same order as the request. Furthermore, not all genes have an associated "presistent_id".

**Comment by mugitty** [comment]
@ayeTown After getting the results from the analysis, the user should be able to narrow down the lists with results that require further investigation. This number is usually very small.

**Comment by ayeTown** [comment]
Okay, thanks for your help!

**Comment by ayeTown** [comment]
@mugitty is there a way to obtain the associated genes (in reference) with each go term? I would like a static data frame of the genes associated with each go term

**Comment by mugitty** [comment]
@ayeTown, The output is large when the gene ids are added for each of the enriched terms. We will let you know if there is another way to obtain the mappings

**Comment by erzakiev** [comment]
> Currently the API does not output the genes associated with each term

By all means, this is the biggest limitation of the API (read deal-breaking) since its inception. I'd like let the PANTHER team know about it, where do I send my feedback?

**Comment by serenalotreck** [comment]
@mugitty this limitation is also breaking for me, I am completely unable to do anything with a GO enrichment done with the API, and I can't do the enrichments manually because there are too many of them.

Do you think the implementation would be relatively straightforward? I've never worked on an API before, but I'm happy to take a crack at it if that seems reasonable.

**Comment by mugitty** [comment]
@serenalotreck , we are currently looking at modifying the API to output mapping information as part of the result set.

**Comment by serenalotreck** [comment]
Great! Do you have an idea of how long it would be before we see those changes?

**Comment by mugitty** [comment]
@serenalotreck, due to the potential size of the output, we are looking at different ways of outputting the data. I will update the issue when I have a clear idea.

**Comment by mugitty** [comment]
The PANTHER API has been modified as follows:

1. PANTHER mapping -/services/oai/pantherdb/geneinfo has been modified to output the mapped genes associated with the PANTHER id
2. PANTHER Tools - Enrichment (Overrepresentation) - /services/oai/pantherdb/enrich/overrep has been modified with option to to output the mapped genes associated with the comparision list or the reference list


FYI, depending on the request, the output from the API can be rather large.  This can result in web server server IO performance issues.  This may result in updates to the API to limit output size.

**Comment by serenalotreck** [comment]
Thanks so much! I'm looking ta the updated documentation now, and I want to make sure I'm interpreting it correctly -- to get the genes associated with each GO term, you would provide `COMP_LIST` for the `mappedInfo` option?

I'm looking at the output for the example in the API, and I'm still not seeing where the associated between genes and Go terms are, I just see a list of genes with no GO terms associated.

Thanks!

**Comment by mugitty** [comment]
@serenalotreck,
Yes, The COMP_LIST option will return the mappings for the comparison list (i.e. the list you are comparing with the reference list).  Sometimes, something from your list may be mapped to multiple genes in PANTHER or a single gene in PANTHER maybe mapped to multiple genes in your list. So the system returns both mapped_ids and mapped_panther_ids.

For example, lets assume you run your analysis  for 3 genes (STK3,STK24,BRCA1) for annotation data set GO Complete Biological Process.  The request will be as follows:
curl -X POST "https://pantherdb.org/services/oai/pantherdb/enrich/overrep?geneInputList=STK3,STK24,BRCA1&organism=9606&annotDataSet=GO%3A0008150&enrichmentTestType=FISHER&correction=FDR&mappedInfo=COMP_LIST"

If you review the output, you will see the following as one of the items in the result:
      {
        "input_list": {
          "mapped_ids": "STK24,STK3",
          "mapped_panther_ids": "HUMAN|HGNC=11403|UniProtKB=Q9Y6E0"
        },
        "number_in_list": 1,
        "fold_enrichment": 686,
        "fdr": 0.8667928126379352,
        "expected": 0.0014577259475218659,
        "number_in_reference": 10,
        "pValue": 0.001457088509679154,
        "term": {
          "id": "GO:0070572",
          "label": "positive regulation of neuron projection regeneration"
        },
        "plus_minus": "+"
      }

The number in list is 1 which is "HUMAN|HGNC=11403|UniProtKB=Q9Y6E0" and it has been annotated to term  "GO:0070572".  However, two ids from your input list are mapped to the term.

**Comment by serenalotreck** [comment]
Ok awesome, thank you!

I did notice that some of my significantly enriched GO terms don't have any mapped IDs with them -- is that because they're significantly underrepresented, and I'd need to request REF_LIST as well?

**Comment by mugitty** [comment]
If the term is underrepresented, you will see '"plus_minus": "-"'

You do not need to run the analysis with the "REF_LIST" option
